react-native/react/actions/login2.js

```diff
@@ -277,25 +163,301 @@ export function logout () {
+          mapStateToProps: state => state.login2.deviceName
+        }
+      }
+    }))
+  }
+}
+
+function askForOtherDeviceType (cb) {
+  return dispatch => {
+    const props = {
+      onSubmit: otherDeviceRole => {
+        cb(otherDeviceRole)
+      }
+    }
+
+    dispatch(routeAppend({
+      parseRoute: {
+        componentAtTop: {
+          component: ExistingDevice,
+          mapStateToProps: state => state.login2.codePage,
+          props
+        }
+      }
+    }))
+  }
+}
+
+function askForCodePage (cb) {
+  return dispatch => {
+    const props = {
+      setCodePageMode: mode => dispatch(setCodePageMode(mode)),
+      qrScanned: code => cb(code.data),
+      setCameraBrokenMode: broken => dispatch(setCameraBrokenMode(broken)),
+      textEntered: text => cb(text),
+      doneRegistering: () => dispatch(doneRegistering())
+    }
+
+    const mapStateToProps = state => {
+      const {
+        mode, codeCountDown, textCode, qrCode,
+        myDeviceRole, otherDeviceRole, cameraBrokenMode } = state.login2.codePage
+      return {
+        mode,
+        codeCountDown,
+        textCode,
+        qrCode,
+        myDeviceRole,
+        otherDeviceRole,
+        cameraBrokenMode
+      }
+    }
+
+    dispatch(routeAppend({
+      parseRoute: {
+        componentAtTop: {
+          component: CodePage,
+          mapStateToProps,
+          props
+        }
+      }
+    }))
+  }
+}
+
+export function registerWithExistingDevice () {
+  return (dispatch, getState) => {
+    const provisionMethod = enums.provisionUi.ProvisionMethod.device
+    startLoginFlow(dispatch, getState, provisionMethod, null, null, Constants.actionRegisteredWithExistingDevice)
+  }
+}
+
+export function registerWithUserPass () {
+  return (dispatch, getState) => {
+    const title = 'Registering with your Keybase passphrase'
+    const subTitle = 'lorem ipsum lorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsum'
+    const provisionMethod = enums.provisionUi.ProvisionMethod.passphrase
+    startLoginFlow(dispatch, getState, provisionMethod, title, subTitle, Constants.actionRegisteredWithUserPass)
+  }
+}
+
+export function registerWithPaperKey () {
+  return (dispatch, getState) => {
+    const title = 'Registering with your paperkey requires your username'
+    const subTitle = 'Different lorem ipsum lorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsumlorem ipsum'
+    const provisionMethod = enums.provisionUi.ProvisionMethod.paperKey
+    startLoginFlow(dispatch, getState, provisionMethod, title, subTitle, Constants.actionRegisteredWithPaperKey)
+  }
+}
+
+function startLoginFlow (dispatch, getState, provisionMethod, userPassTitle, userPassSubtitle, successType) {
+  const mobile = true // TODO desktop also
+  const deviceType = mobile ? 'mobile' : 'desktop'
+  const incomingMap = makeKex2IncomingMap(dispatch, getState, provisionMethod, userPassTitle, userPassSubtitle)
```

Comment:
Reasonable to accept `...incomingMapArgs` above and spread them here, instead of repeating all of the arg names?